include/arch/arm/aarch64/arm_mmu.h

```diff
@@ -33,25 +33,39 @@
  * current usage of attrs[31:3] is:
  * attrs[3] : Access Permissions
  * attrs[4] : Memory access from secure/ns state
- * attrs[5] : Execute Permissions
+ * attrs[5] : Execute Permissions privileged mode (PXN)
+ * attrs[7] : Execute Permissions unprivileged mode (UXN)
+ * attrs[6] : Mirror RO/RW permissions to EL0
  *
  */
 #define MT_PERM_SHIFT		3U
 #define MT_SEC_SHIFT		4U
-#define MT_EXECUTE_SHIFT	5U
+#define MT_P_EXECUTE_SHIFT	5U
+#define MT_U_EXECUTE_SHIFT	6U
+#define MT_RW_EL0h_SHIFT	7U
 
 #define MT_RO			(0U << MT_PERM_SHIFT)
 #define MT_RW			(1U << MT_PERM_SHIFT)
 
+#define MT_RW_EL0h		(1U << MT_RW_EL0h_SHIFT)
+#define MT_RW_ELh		(0U << MT_RW_EL0h_SHIFT)
+
 #define MT_SECURE		(0U << MT_SEC_SHIFT)
 #define MT_NS			(1U << MT_SEC_SHIFT)
 
-#define MT_EXECUTE		(0U << MT_EXECUTE_SHIFT)
-#define MT_EXECUTE_NEVER	(1U << MT_EXECUTE_SHIFT)
+#define MT_P_EXECUTE		(0U << MT_P_EXECUTE_SHIFT)
+#define MT_P_EXECUTE_NEVER	(1U << MT_P_EXECUTE_SHIFT)
+
+#define MT_U_EXECUTE		(0U << MT_U_EXECUTE_SHIFT)
+#define MT_U_EXECUTE_NEVER	(1U << MT_U_EXECUTE_SHIFT)
 
-/* Some compound attributes for most common usages */
-#define MT_CODE			(MT_NORMAL | MT_RO | MT_EXECUTE)
-#define MT_RODATA		(MT_NORMAL | MT_RO | MT_EXECUTE_NEVER)
+#define MT_P_RW_U_RW		(MT_RW | MT_RW_EL0h | MT_P_EXECUTE_NEVER | MT_U_EXECUTE_NEVER)
+#define MT_P_RW_U_NA		(MT_RW | MT_RW_ELh  | MT_P_EXECUTE_NEVER | MT_U_EXECUTE_NEVER)
+#define MT_P_RO_U_RO		(MT_RO | MT_RW_EL0h | MT_P_EXECUTE_NEVER | MT_U_EXECUTE_NEVER)
+#define MT_P_RO_U_NA		(MT_RO | MT_RW_ELh  | MT_P_EXECUTE_NEVER | MT_U_EXECUTE_NEVER)
+#define MT_P_RWX_U_RWX		(MT_RW | MT_RW_EL0h | MT_P_EXECUTE | MT_U_EXECUTE)
```

Comment:
MT_P_RWX_U_RWX - looking at current code, this combination will not be possible as we have marked RW memory as execute never, be it privileged/unprivileged mode . For kernel mode, motivation to mark RW as Execute Never was to prevent execution from data region.

Reply:
ACK